### ISSUE TITLE
python: deadlock when importing modules that register tasks

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#537]
We are observing a deadlock with python processors when running on multiple nodes that goes like this:
* thread 1 on the python processor is running a task (the top level task?) that imports a python module.
 * as part of that import, legion tasks are registered, which causes a task registration to be queued up on the local python processor, and then that task waits on an event related to inter-node communication in the legion code
 * the thread scheduler sees that thread 1 wants to wait on an event, so it suspends that thread and spins up a new thread (thread 2) to try to use useful python processor work
 * thread 2 sees that there are pending task registrations, so starts to work on the first one of those.
 * part of task registration is importing the python module (I don't think we test that it's already present), and that import blocks on the python interpreter's module import mutex (https://docs.python.org/2.7/library/threading.html#importing-in-threaded-code).
 * this block is not done with realm's knowledge, so now we have thread 2 waiting for thread 1 to finish its import, and thread 1 waiting for thread 2 to finish or sleep so that it can resume execution and finish its import

There are a number of possible things we could consider doing to address this, but it's likely they all have drawbacks, so let's discuss them here:
 1. Eliminate the blocking wait in the Legion task registration path so that thread 1 never gets suspended [@lightsighter to explain the drawbacks here]
 1. Change how task registration works in python bindings to defer the actual registration until any imports are complete. [@elliottslaughter to comment here]
 1. Somehow poke into the python interpreter and understand if the import lock is held by a task that is about to suspend, and if it is, spinwait instead.
 1. Change the execution model of a Realm `LocalPythonProcessor` to allow tasks to run concurrently.  Any time a python task suspends and leaves zero active workers, another worker is started as before.  The difference is that blocked workers would be resumed immediately rather than waiting for the active worker to complete or suspend.  This would complicate the performance model for python tasks (i.e. you wouldn't be able to count on exclusive access to an execution resource).
 1. Create a "scheduler lock" for a Realm `LocalPythonProcessor` and expose it as part of the python bindings.  Python code that takes any python-interpreter-implemented mutex could be wrapped in a `with realm_python_scheduler_lock:`, causing that task to spinwait rather than yield the execution resource if/when it waits on events.
 1. `@legion.task` queuing up tasks and using a realm callback to complete them at the end of the current task execution (suggested below)

**Comment by lightsighter** [comment]
As to solution 1, that's not the only place this happens. @elliottslaughter can also cause the hang doing any number of other things such as setting semantic values or registering any number of other kinds of resources with the runtime. I also think it's even worse than that, pretty much anywhere that the runtime takes a Realm fast reservation could potentially result in a preemption of the thread and therefore a hang. It's going to be very difficult to make all the potential paths that the python interpreter could call at import time lock-free.

I've always disliked that Python could call into the Legion runtime after starting the runtime but not inside of any legion tasks. I feel like the best answer in this case is solution 2, that modules just need to track what they've registered and only register it lazily on demand inside of actual tasks and not inside of their import statements.

**Comment by elliottslaughter** [comment]
I think some clarifications are in order.

The scenario in which we're freezing is one where we are using DCR on the top-level task `legion.legion_main`. This is a built-in task whose job it is to call `exec` on the actual, user-defined script. The whole goal of this process is to make it appear like a normal Python execution (i.e. how `python user.py` would behave), where the script-style execution of `user.py` is DCR'd.

This has several implications:

 a. This is *not* happening in the "started-but-not" uncanny valley that plagued us before. The top-level task has started.

 b. The module for `user.py` is not a real module and therefore cannot be imported normally. The module is called `__legion_main__` (following the convention on Python's `__main__`) and you can only look it up in `sys.modules`; you cannot just do a raw import on it. I assume Python's import logic has some smarts to deal with this, because otherwise things would be even more wildly broken than they are now, just something to keep in mind.

 c. The more important implication is that *we must block in task registration*. To do otherwise would mean that some shards in the DCR'd top-level task could run ahead of others and get a task launch through, and in theory the mapping and execution of those tasks could race with registration on other nodes.

Point (c) reminded me that I was supposed to add a barrier to block execution of the user script around task registration. However, an initial patch to do that makes the situation worse so I think at the moment I'm just going to ignore this aspect. I will at least put the patch in a branch so we have it.

I think this means (1) can't work for us. (2) is similarly not a solution; we're in the top-level task. I don't know about (3) through (5).

**Comment by lightsighter** [comment]
I still think (2) is a solution. You can still defer the runtime calls from inside the import statement until you get another call into the library at which point you check for any deferred runtime calls and then perform them.

**Comment by elliottslaughter** [comment]
Here's the code we're running. We don't call import, and we don't hold any locks:

https://github.com/StanfordLegion/legion/blob/master/bindings/python/legion.py#L1527-L1537

Therefore I don't see how (2) is applicable.

**Comment by lightsighter** [comment]
Ok, so how did we get into import code? We're definitely inside an import statement when we hang.

**Comment by elliottslaughter** [comment]
I guess what must be happening is that `user.py` is importing another module, and that module is transitively registering a task.

I'm still confused about (2) though. We're not running in a mode where there is a clear boundary between "all imports" and "all execution"; these are interleaved. Unless I can override Python's import mechanism or add a callback that occurs when an import completes, I don't see how to do this.

Also, I'm not sure this is likely to be an issue, but it's worth mentioning because it will violate Python's normal semantics: it will be impossible to invoke any tasks at the top-level of a module.

**Comment by elliottslaughter** [comment]
Ok, here's a proposal:

We could add a new callback in the Realm Python module. When loading `legion.py` this would be auto-populated with a Python function that would handle the deferred registration of tasks. Here are the steps that would be followed during task registration:

  * From user code: define a task with `@task`
      * `legion.task` puts the user's task into a list to be processed later
  * From the Realm Python processor main loop: calls `legion.py` registration callback
      * The registration callback completes any registrations (via `register_task_variant`)
      * The Realm Python processor main loop then receives the task registrations and processes them as normal
      * Note that for safety, the Realm Python Processor has to call the `legion.py` registration callback *again*, because if any imports occurred there could be more tasks waiting to be registered
      * Repeat until convergence
      * Finally call the next task in the queue

**Comment by lightsighter** [comment]
I guess the two questions are how does Realm get the call-back registered and how often does Realm have to invoke this call-back?

**Comment by streichler** [comment]
@lightsighter , you said above that there are other things a legion task can do that will cause blocking waits.  If so, then deferring only the task registrations (i.e. #2 or the soon-to-be-copied-up #6) wouldn't be sufficient, would it?

**Comment by elliottslaughter** [comment]
Actually, wait, this is going to run into exactly the same freeze we're already hitting. Because Realm spins up a new thread whenever the Python processor is blocked, we're just going to call the callback and cause the freeze. The real root of the condition is holding the import lock while blocking in Realm, and my approach doesn't fix that.

Is there a reason why we couldn't do something like: processes deferred task registrations *only* if there are > 0 tasks in the queue? That way we'll still spin up the thread, but it won't actually attempt to grab the import lock unless there is work in the queue, which we can avoid as long as you don't call any tasks at the top level of a module import. That sure sounds like a footgun, but it was exactly the same footgun I was proposing before, so we aren't actually losing any robustness in that respect.

**Comment by lightsighter** [comment]
Yes, all runtime calls need to be deferred, not just task variant registration. Anything that can call into the runtime can potentially attempt to take a Realm fast reservation and block the thread if there is contention.

**Comment by streichler** [comment]
Ok, the current consensus is to do proposal #5 (the scheduler lock), along with some monkeypatching of the Python bindings for the Legion API to wrap the various registration tasks (@lightsighter to provide a list) with automatic locking of the scheduler for the duration of that call.

**Comment by lightsighter** [comment]
Here is a first pass on a list of API calls that I think would be called inside an import statement and could potentially cause bad behavior:
MPI Interop: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6089-6118
Semantic Values: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6120-6423
Printing: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6426-6446
Mapper Registration: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6448-6514
Projection and Sharding: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6517-6619
Layout Constraint Registration: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6988-7045
Task Variant Registration: https://gitlab.com/StanfordLegion/legion/blob/master/runtime/legion.h#L6988-7045

**Comment by streichler** [comment]
I think the burst of commits ending in c2f4c9d takes care of this.  @elliottslaughter , can you think of anything that still needs doing here?

**Comment by elliottslaughter** [comment]
I need to check whether the workaround at https://upc-bugs.lbl.gov/bugzilla/show_bug.cgi?id=3908 is still necessary, but otherwise I think we're good at the moment.

**Comment by streichler** [comment]
I don't think we should remove that workaround.  The problem with forks occurring while active messages are being sent is real, and it may be that changes to the python startup cause it to stop reproducing the fork issue, giving us a false sense of security.
